File: project/suchan-onboarding/application/src/main/java/com/survey/application/exception/GlobalExceptionHandler.java
package com.survey.application.exception;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@ControllerAdvice
public class GlobalExceptionHandler {
    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ErrorResponse> handleValidationExceptions(IllegalArgumentException ex) {
        return new ResponseEntity<>(new ErrorResponse(ex.getMessage()), HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleValidationExceptions(MethodArgumentNotValidException ex) {
        Map<String, String> errors = new HashMap<>();

        bindFieldError(ex, errors);

        bindObjectError(ex, errors);

        logger.error("validation error : {}", errors);
        return new ResponseEntity<>(new ErrorResponse("validation error", errors), HttpStatus.BAD_REQUEST);
    }

    private void bindFieldError(MethodArgumentNotValidException ex, Map<String, String> errors) {
        ex.getBindingResult().getFieldErrors().forEach((error) -> {
            String fieldName = error.getField();
            String errorMessage = error.getDefaultMessage();
            errors.put(fieldName, errorMessage);
        });
    }

    private void bindObjectError(MethodArgumentNotValidException ex, Map<String, String> errors) {
        ex.getBindingResult().getGlobalErrors().forEach((error) -> {
            String objectName = error.getObjectName();
            String errorMessage = error.getDefaultMessage();
            errors.put(objectName, errorMessage);
        });
    }

    record ErrorResponse(String errorMessage, Map<String, String> errors, LocalDateTime timestamp) {
        ErrorResponse(String errorMessage) {
            this(errorMessage, null, LocalDateTime.now());
        }

        ErrorResponse(String errorMessage, Map<String, String> errors) {
            this(errorMessage, errors, LocalDateTime.now());
        }
    }
}

File: project/koodongha-onboarding/api/src/main/kotlin/com/example/exception/GlobalExceptionHandler.kt
package com.example.exception

import com.example.exception.*
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import org.springframework.web.bind.MethodArgumentNotValidException

@ControllerAdvice
class GlobalExceptionHandler {

    private val log: Logger = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    @ExceptionHandler(BusinessException::class)
    fun handleBusinessException(e: BusinessException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        log.warn("BusinessException 발생: code=${e.errorCode}, message=${e.message}")
        val response = ErrorResponse.of(e.errorCode, e.message, request.requestURI)
        return ResponseEntity.status(e.errorCode.status).body(response)
    }

    @ExceptionHandler(IllegalArgumentException::class)
    fun handleIllegalArgument(e: IllegalArgumentException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        log.warn("잘못된 요청 파라미터: ${e.message}")
        val response = ErrorResponse.of(ErrorCode.INVALID_INPUT, e.message, request.requestURI)
        return ResponseEntity.badRequest().body(response)
    }

    @ExceptionHandler(MethodArgumentTypeMismatchException::class)
    fun handleTypeMismatch(e: MethodArgumentTypeMismatchException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        log.warn("파라미터 타입 불일치: ${e.message}")
        val response = ErrorResponse.of(ErrorCode.INVALID_INPUT, e.message, request.requestURI)
        return ResponseEntity.badRequest().body(response)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(e: MethodArgumentNotValidException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        val errorMessage = e.bindingResult.fieldErrors
            .joinToString(", ") { "${it.field}: ${it.defaultMessage}" }
        log.warn("요청 값 검증 실패: $errorMessage")
        val response = ErrorResponse.of(ErrorCode.INVALID_INPUT, errorMessage, request.requestURI)
        return ResponseEntity.badRequest().body(response)
    }

    @ExceptionHandler(Exception::class)
    fun handleGenericException(e: Exception, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        log.error("Unhandled Exception 발생!", e)
        val response = ErrorResponse.of(ErrorCode.INTERNAL_ERROR, e.message, request.requestURI)
        return ResponseEntity.status(ErrorCode.INTERNAL_ERROR.status).body(response)
    }
}
